### DISCUSSION TITLE
[ 功能建议 ] DDL/DML独立审批流程审核

### DISCUSSION THREAD

**Discussion by wineejam** [hhyo/Archery#1607] [General]
由于公司业务特殊，需要DDL、DML分开审批，于是我修改了部分代码实现了此效果，有需要的可以自行下载

[Archery-1.8.0_patch_ddl_dml.zip](https://github.com/hhyo/Archery/files/7623690/Archery-1.8.0_patch_ddl_dml.zip)

具体操作步骤请看 readme.md
### 

## 注意事项：

- 此patch包只在1.8.0版本测试过，其它版本不保证
- 操作前请自行备份文件，patch.sh 会直接覆盖掉原文件

### 2. 解压并查看文件目录结构，对比是否和以下结构一致

```bash
patch_ddl_dml
├── common
│   ├── static
│   │   └── ace
│   │       └── ace-init.js
│   ├── templates
│   │   └── config.html
│   └── utils
│       └── sendmsg.py
├── patch_for_1.8.0.sql
├── patch.sh
├── readme.md
├── sql
│   ├── archiver.py
│   ├── models.py
│   ├── query_privileges.py
│   ├── resource_group.py
│   ├── sql_workflow.py
│   ├── templates
│   │   ├── archive.html
│   │   ├── queryapplylist.html
│   │   └── sqlsubmit.html
│   └── utils
│       └── workflow_audit.py
└── static
    └── ace
        └── ace-init.js

10 directories, 16 files
```

### 3. 导入SQL文件

```sql
ALTER TABLE workflow_audit_setting ADD sql_type int(11) NOT NULL DEFAULT '0' COMMENT 'SQL类型，1：DDL 2：DML 0：other'; 

ALTER TABLE workflow_audit_setting DROP INDEX workflow_audit_setting_group_id_workflow_type_5884053a_uniq;

ALTER TABLE workflow_audit_setting ADD UNIQUE index `workflow_audit_setting_group_id_workflow_type_5884053a_uniq` (`group_id`,`workflow_type`,`sql_type`) ;
```

### 4. 运行patch.sh 覆盖文件

```bash
cd archery_root_path/patch_ddl_dml
sh patch.sh
```

### 5. 重启Archery

```bash
cd archery_root_path/
sh admin.sh restart
```

**Comment by jimsoft1000** [comment]
如果DDL和DML写在一起，你会校验吗

**Comment by wineejam** [comment]
> 如果DDL和DML写在一起，你会校验吗

同时存在DDL和DML的工单。以DDL审批流程为准

**Comment by czxin788** [comment]
我在前端做了ddl和dml校验，ddl工单不能写dml语句

![image](https://user-images.githubusercontent.com/13861904/141264630-60bb80a3-5fe4-42e1-a46f-4fdfe009446a.png)

**Comment by wineejam** [comment]
> 我在前端做了ddl和dml校验，ddl工单不能写dml语句
> 
> ![image](https://user-images.githubusercontent.com/13861904/141264630-60bb80a3-5fe4-42e1-a46f-4fdfe009446a.png)

目前我的做法是不需要选择DDL还是DML。会根据sql内容自动判断。然后自动选不同的审核流程

**Comment by shuaishuai8147** [comment]
DDL那里可以选择吗，是下拉框吗
